Sources/NIOSSL/SSLContext.swift

```diff
@@ -520,6 +527,83 @@ extension NIOSSLContext {
+            }
+            
+            // We want to take the SSL pointer and extract the parent Swift object. These force-unwraps are for
+            // safety: a correct NIO program can never fail to set these pointers, and if it does failing loudly is
+            // more useful than failing quietly.
+            let parentCtx = CNIOBoringSSL_SSL_get_SSL_CTX(ssl)!
+            let parentPtr = CNIOBoringSSLShims_SSL_CTX_get_app_data(parentCtx)!
+            let parentSwiftContext: NIOSSLContext = Unmanaged.fromOpaque(parentPtr).takeUnretainedValue()
+
+            // Similarly, this force-unwrap is safe because a correct NIO program can never fail to unwrap this entry
+            // either.
+            guard let userChosenContext = parentSwiftContext.sniManager!.sniInicated(hostname: serverNameString, context: parentSwiftContext) else
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // The user may or may not retain a reference to the NIOSSLContext
+            // that they return here, if they provide one. The NIOSSLContext finalizer
+            // explicitly decrements the reference count on the underlying SSL_CTX.
+            //
+            // However, SSL_set_SSL_CTX explicitly increases the reference count and
+            // will decrement the count when the SSL* object is destroyed, thus finally
+            // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+            //
+            // As such, no memory should be leaked or freed too early in the event that
+            // the user fails to retain their created and return NIOSSLContext. In the
+            // event that they do retain it, the reference count will remain at at-least
+            // 1 and thus they can re-use the same context for multiple connections.
+            
+            guard let nativeReturnVal = CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext) else
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+            // told to assign on success. On error, this will not happen, so we check for equality
+            // here and handle any such errors by signaling the connection to fail.
+            if nativeReturnVal != userChosenContext.sslContext
+            {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+            
+            // If no action was taken, we indicate that this current context can be used
+            // to attempt to complete the handshake. Given that the user has indicated
+            // that they want control of setting the context when SNI is indicated, if this
+            // indication turns out to be false, the SSL/TLS connection will fail with an
+            // appropriate error and the user will be made aware of it.
+            return SSL_TLSEXT_ERR_OK
```

Comment:
I don't think we're quite done here. The `SSLConnection` object itself has a reference to its parent context, which is now different. We need to update that. We also need to update any state derived from it, such as the call to `setRenegotiationSupport` in `SSLConnection.init`. It would be good to add unit tests that validate that any change to that field is correctly reflected when we change the context.

Reply:
Good point regarding updating the native state object. However, I don't want to get too crazy with testing here beyond making sure the linked swift objects change because then we're venturing into testing the correctness of the native crypto library itself. 

To resolve any possible confusion, what's actually happening here is we're changing the `SSL_CTX` for **this single client connection alone**, represented by the `SSL*` parameter that BoringSSL is feeding us in this callback.

The original server `SSLContext`, the one bound to the `TlsConfiguration` that has the SNI callback set on it, never goes away or changes. It remains as the default context for the stream wrapping the TCP listener on the server.

Something like this pseudo:

```
defaultSslCtx = new sslctx + sni callback
let clientConn = new sslstream<clientTcpConn, defaultSslCtx>() from tcpListener.accept
clientConn.handshake
snicallback invoked, SSL* == clientConn
create brand new context with entirely tailored settings, crafted for the given host
clientConn.context = brand new context // original context remains in use for all new client connections
```

So yeah I think for sure it's a great call to verify that the relationships between the swift objects and the native objects are being correctly maintained. But if we get into doing things like verifying that BoringSSL has actually done its job then I think we're leaving our boundary and overlapping into testing the native library itself.

Reply:
:+1: I think that's entirely reasonable here. Because of the limited type visibility I think the best way to do this is to use the only thing we know changes, which is the renegotiation support. If we add a single test to validate this, say by twiddling from a context that _does_ support renegotiation to one that doesn't, it should be sufficient to prove we've done the right thing I think.